### PR TITLE
Production decomposition

### PR DESCRIPTION
**Pull request by isc-pbarton**
Added a new configurable setting: "Decompose Productions". If enabled, interoperability productions will be represented in the source control repository as a collection of files for individual item settings and production settings. The goal is to allow multiple users in a shared development environment to edit the same production without overwriting or blocking each others' changes.

### CODE REVIEW

File: cls/SourceControl/Git/PullEventHandler/IncrementalLoad.cls [hunk 1]
```diff
@@ -38,29 +38,45 @@ Method OnPull() As %Status
...
+        1: ##class(SourceControl.Git.Utils).Type(item)
+    )
     set name = ##class(SourceControl.Git.Utils).NameWithoutExtension(item)
+    set settings = ##class(SourceControl.Git.Settings).%New()
     set deleted = 1
     if type = "prj" {
         set sc = $system.OBJ.DeleteProject(name)
     }elseif type = "cls" {
-        set sc = $system.OBJ.Delete(item)
+        if settings.decomposeProductions && ##class(SourceControl.Git.Production).IsProductionClass(name, "FullExternalName") {
```
Comment (isc-pbarton) [inline cls/SourceControl/Git/PullEventHandler/IncrementalLoad.cls:54]: Use ItemIsProductionToDecompose utility method

File: cls/SourceControl/Git/Utils.cls [hunk 2]
```diff
@@ -1252,8 +1260,32 @@ ClassMethod ImportItem(InternalName As %String, force As %Boolean = 0, verbose A
...
+                if (targetProduction '= "") && '$$$comClassDefined(targetProduction) {
+                    set sc = ##class(SourceControl.Git.Production).CreateProduction(targetProduction)
+                }
+                if $$$ISOK(sc) {
+                    set rollbackFile = ##class(%File).TempFilename()
+                    set sc = ##class(Ens.Deployment.Deploy).DeployCode(filename,targetProduction,0,rollbackFile)
+                    do ##class(%File).Delete(rollbackFile)
+                }
+            }
+        } elseif (type = "cls") && settings.decomposeProductions
```
Comment (isc-pbarton) [inline cls/SourceControl/Git/Utils.cls:1283]: Use ItemIsProductionToDecompose()

File: cls/SourceControl/Git/Production.cls [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+        set rs = ##class(%SQL.Statement).%ExecDirect(
+            ,"select Name, ClassName from Ens_Config.Item where Production = ?"
+            , productionName)
+        $$$ThrowSQLIfError(rs.%SQLCODE, rs.%Message)
+        while rs.%Next() {
+            if '$get(^mtempsscProd($job,"items", $listbuild(rs.Name, rs.ClassName))) {
+                set itemInternalName = ..CreateInternalName(productionName, rs.Name, rs.ClassName, 0)
+                set modifiedItems(itemInternalName) = "A"
+            }
+            kill ^mtempsscProd($job,"items", $listbuild(rs.Name, rs.ClassName))
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Production.cls:203]: Style: would be better to use ^IRIS.Temp these days

File: cls/SourceControl/Git/Production.cls [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+{
+    set sc = $$$OK
+    set productionName = $replace($piece($replace(externalName,"\","/"),"/",*-1),"_",".")
+    set production = ##class(Ens.Config.Production).%OpenId(productionName,,.sc)
+    $$$QuitOnError(sc)
+    set itemToRemove = $$$NULLOREF
+    for i=1:1:production.Items.Count() {
+        set configItem = production.Items.GetAt(i)
+        set itemInternalName = ..CreateInternalName(productionName, configItem.Name, configItem.ClassName)
+        set itemExternalName = $classmethod($$SrcCtrlCls^%buildccr, nameMethod, itemInternalName)
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Production.cls:317]: Better to use ##class(%Studio.SourceControl.Interface).SourceControlClassGet() - maybe wrap in a macro

File: cls/SourceControl/Git/Settings.cls [hunk 5]
```diff
@@ -23,6 +23,9 @@ Property percentClassReplace As %String [ InitialExpression = {##class(SourceCon
 /// Git project settings are read-only in the web user interface
 Property settingsUIReadOnly As %Boolean [ InitialExpression = {##class(SourceControl.Git.Utils).SettingsUIReadOnly()} ];
 
+/// Interoperability productions are source-controlled under separate files for each configuration item
+Property decomposeProductions As %Boolean [ InitialExpression = {##class(SourceControl.Git.Utils).DecomposeProductions()} ];
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings.cls:27]: I think this needs UI too

**Review by isc-tleavitt — CHANGES_REQUESTED** [review]
Left a few comments outside the review accidentally - sorry about that. A few initial points of feedback here.

**Comment by isc-pbarton** [comment]
Some additional points that came out of testing today:

- [ ] enhancement: automatically add mapping for /ptd items when turning on production decomposition (nice to have)
- [ ] fix: if I try editing a production item that has been modified by another user, the alert includes a full stack trace
- [ ] fix: in the Git WebUI workspace view, production items changed by another user say "file changed by: undefined" on hover instead of the other user's name.
- [ ] enhancement: in the Git WebUI workspace view, show the internal name of the production item when hovering over an item. This will make it easier to tell which production items have been changed.
- [ ] fix: if source control is enabled but local git repo folder is not set or is invalid, it's impossible to configure because a failure to export the settings file will rollback the entire transaction. The workaround is to set the local git repo folder in the ^SYS global directly.
- [ ] fix: There's a failure to clone an existing repo when running Configure() for the first time because git-source-control.json gets exported before the clone, and you cannot clone into a non-empty directory.
- [ ] fix: git-source-control settings do not import properly when pulling change from remote

File: cls/SourceControl/Git/Production.cls
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Production.cls:1]: @isc-pbarton you've been keeping this synced with %Studio.SourceControl.Production, right? (Assuming yes and skipping review since I've already reviewed that.)

File: cls/SourceControl/Git/Settings.cls [hunk 6]
```diff
@@ -75,6 +81,36 @@ Method %OnNew() As %Status
     quit $$$OK
 }
 
+Method SaveWithSourceControl() As %Status
+{
+    set sc = $$$OK
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings.cls:86]: Important! This method needs to new %SourceControl before calling SourceControlCreate.

File: cls/SourceControl/Git/Settings.cls [hunk 7]
```diff
@@ -75,6 +81,36 @@ Method %OnNew() As %Status
...
+    if 'skipSourceControl {
+        // Source control settings naively by only calling OnAfterSave hooks
+        // Future enhancement: add full source control support with settings UI
+        if (##class(%Studio.SourceControl.Interface).SourceControlClassGet() = ##class(SourceControl.Git.Extension).%ClassName(1)) {
+            if '##class(SourceControl.Git.Utils).IsInSourceControl(internalName) {
+                set sc = ##class(SourceControl.Git.Utils).AddToSourceControl(internalName)
+                $$$QuitOnError(sc)
+            }
+        }
+        if ($IsObject(%SourceControl)) {
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings.cls:107]: In general, use $IsObject($Get(%SourceControl))

File: cls/SourceControl/Git/Settings.cls [hunk 8]
```diff
@@ -75,6 +81,36 @@ Method %OnNew() As %Status
...
+Method SaveWithSourceControl() As %Status
+{
+    set sc = $$$OK
+    #dim %SourceControl As %Studio.SourceControl.Interface
+    if '$isobject($get(%SourceControl)) {
+        do ##class(%Studio.SourceControl.Interface).SourceControlCreate($username)
+    }
+    set gitDir = ##class(%File).NormalizeDirectory(..namespaceTemp)_".git"
+    set skipSourceControl = ('##class(%File).DirectoryExists(gitDir) && 
+        (##class(%Studio.SourceControl.Interface).SourceControlClassGet() = ##class(SourceControl.Git.Extension).%ClassName(1)))
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings.cls:93]: This is a bit tricky because we could use a composite extension like in IPM or CodeTidy - that is, we could be using Embedded Git with the source control class set to one other than SourceControl.Git.Extension.

File: cls/SourceControl/Git/Settings.cls [hunk 9]
```diff
@@ -133,6 +169,12 @@ Method %Save() As %Status
     // update value of basicUserMode to reflect the updated setting for basicMode
     set ..userBasicMode = ##class(SourceControl.Git.Utils).UserBasicMode()
     
+    set @storage@("settings","decomposeProductions") = ..decomposeProductions
+    if (..decomposeProductions & ($data(..Mappings("PTD"))<10)) {
+        set ..Mappings("PTD","*") = "ptd/" // configure a default mapping for PTD items
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings.cls:174]: Question: do we allow a mapping other than "ptd" for these? Or is the folder hardcoded anywhere?

File: cls/SourceControl/Git/Settings.cls [hunk 10]
```diff
@@ -141,6 +183,51 @@ Method %Save() As %Status
...
+    kill ..Mappings
+    set mappingsDyn = pSettingsDyn.%Get("Mappings", {})
+    set i1 = mappingsDyn.%GetIterator()
+    while i1.%GetNext(.k1, .v1) {
+        set i2 = v1.%GetIterator()
+        while i2.%GetNext(.k2, .v2) {
+            set ..Mappings(k1, k2) = v2.%Get("directory")
+            if v2.%Get("noFolders") {
+                set ..Mappings(k1, k2, "NoFolders") = 1
+            }
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings.cls:226]: This needs an else to clear the flag if NoFolders changes - otherwise we could get stuck at 1

File: cls/SourceControl/Git/Settings/Document.cls [hunk 11]
```diff
@@ -0,0 +1,108 @@
+/// Custom studio document type for git-source-control settings that are controlled by a file
+Class SourceControl.Git.Settings.Document Extends %Studio.AbstractDocument
+{
+
+Projection RegisterExtension As %Projection.StudioDocument(DocumentExtension = "GSC", DocumentNew = 0, DocumentType = "json");
+
+Parameter INTERNALNAME = "git-source-control.GSC";
+
+Parameter EXTERNALNAME = "git-source-control.json";
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings/Document.cls:9]: Can we call this "embedded-git-config.json" instead?

File: cls/SourceControl/Git/Settings/Document.cls [hunk 12]
```diff
@@ -0,0 +1,108 @@
+/// Custom studio document type for git-source-control settings that are controlled by a file
+Class SourceControl.Git.Settings.Document Extends %Studio.AbstractDocument
+{
+
+Projection RegisterExtension As %Projection.StudioDocument(DocumentExtension = "GSC", DocumentNew = 0, DocumentType = "json");
+
+Parameter INTERNALNAME = "git-source-control.GSC";
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Settings/Document.cls:7]: Can we call this "embedded-git-config.gsc"?

File: cls/SourceControl/Git/Utils.cls [hunk 13]
```diff
@@ -2186,12 +2261,6 @@ ClassMethod GitStatus(ByRef files, IncludeAllFiles = 0)
     }
 }
 
-ClassMethod EmptyInitialCommit()
```
Comment (isc-tleavitt) [inline cls/SourceControl/Git/Utils.cls:2189]: Why was this removed?

**Review by isc-tleavitt — CHANGES_REQUESTED** [review]
A few questions, a few changes requested.

File: cls/SourceControl/Git/Production.cls
Comment (isc-pbarton) [reply cls/SourceControl/Git/Production.cls:1]: This version is slightly out of date with the %Studio version. I will sync them back up and test again.

File: cls/SourceControl/Git/Settings.cls (hunk 9 above)
Comment (isc-pbarton) [reply cls/SourceControl/Git/Settings.cls:174]: I've been testing with the PTD items mapped to a different subdirectory and all is working as expected.

File: cls/SourceControl/Git/Utils.cls (hunk 13 above)
Comment (isc-pbarton) [reply cls/SourceControl/Git/Utils.cls:2189]: This was removed because, instead of doing an empty initial commit, we now export the git-source-control.json file (soon to be renamed) and include that in the initial commit.

**Review by isc-tleavitt — APPROVED** [review]
